Userland/Libraries/LibVirtGPU/CommandBufferBuilder.cpp

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2022, Sahan Fernando <sahan.h.fernando@gmail.com>
+ * Copyright (c) 2022, Stephan Unverwerth <s.unverwerth@serenityos.org>
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/StringView.h>
+#include <Kernel/API/VirGL.h>
+#include <sys/ioctl_numbers.h>
+
+#include <LibVirtGPU/CommandBufferBuilder.h>
+#include <LibVirtGPU/VirGLProtocol.h>
+
+namespace VirtGPU {
+
+static u32 encode_command(u32 length, u32 mid, Protocol::VirGLCommand command)
+{
+    u32 command_value = to_underlying(command);
+    return (length << 16) | ((mid & 0xff) << 8) | (command_value & 0xff);
```

Comment:
`mid` could be clearer as a `u8` argument if we always `& 0xff` it.

Reply:
@ccapitalK Can you help out? I think mid is the command id or something. It's really hard to find information about this protocol :sweat_smile: best I could find is this: https://github.com/freedesktop/virglrenderer/blob/master/src/vrend_decode.c#L1903

Reply:
Mid probably would be clearer as a u8. It isn't the command id (that's specified by command value), its interpretation is usually context dependent, usually used for things like specifying the object type in a bind command.

Reply:
I rewrite this a bit so it's clear what the mid parameter does. Also added some checks so we don't accidentally cause a u16 overflow.

Reply:
Ooh this is much better! Thanks!